src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseStringContainsCharOverloadWithSingleCharactersAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class UseStringContainsCharOverloadWithSingleCharactersAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string CA1839 = "CA1839";
+        private static readonly LocalizableString s_localizableTitle_CA1839 = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseStringContainsCharOverloadWithSingleCharactersTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage_CA1839 = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseStringContainsCharOverloadWithSingleCharactersMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString? s_localizableDescription_CA1839 = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseStringContainsCharOverloadWithSingleCharactersDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor s_rule_CA1839 = DiagnosticDescriptorHelper.Create(
+            CA1839,
+            s_localizableTitle_CA1839,
+            s_localizableMessage_CA1839,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableDescription_CA1839,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule_CA1839);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(CheckIfRuleIsApplicableAndRegister);
+        }
+
+        private void CheckIfRuleIsApplicableAndRegister(CompilationStartAnalysisContext ctx)
+        {
+            var stringType = ctx.Compilation.GetSpecialType(SpecialType.System_String);
+            var charType = ctx.Compilation.GetSpecialType(SpecialType.System_Char);
+            var stringMembers = stringType.GetMembers();
+            var ruleIsApplicable = stringType.GetMembers("Contains").OfType<IMethodSymbol>().Any(m =>
+            {
+                var firstArgument = m.Parameters.First();
+                if (firstArgument is null)
+                    return false;
+
+                return firstArgument.Type.Equals(charType);
```

Comment:
Symbol comparisons should use SymbolEqualityComparer

Reply:
Isnt it safe to call the regular `.Equals` if the Symbol Interface (`IMethodSymbol` in this case) implements `IEquatable<ISymbol>` or am i completely off?

Reply:
Using Equals without specifying a SymbolEqualityComparer is usually okay. But after the introduction of nullable reference types, SymbolEqualityComparer was added so the intent is clear on whether the developer wants to include nullability or not.

There is a rule (RS1024) that flags calls without SymbolEqualityComparer, it's unfortunately currently disabled due to a codefix crash.

Reply:
I was wondering what that was about - thanks for the explanation 😊